Added two examples - One to configure NAT instance for Internet access from private instances, another for a DNS VM to allow on-premises network to query DNS hostnames in the VCN

sivaedupuganti-oracle opened the pull request:
Added example to configure NAT instance for Internet access from private instances

briangustafson commented on docs/examples/networking/nat/vcn.tf:
```diff
@@ -0,0 +1,215 @@
+    user_ocid = "${var.user_ocid}"
+    fingerprint = "${var.fingerprint}"
+    private_key_path = "${var.private_key_path}"
+    private_key_password = "${var.private_key_password}"
```
Need to define this as a variable.

briangustafson commented on docs/examples/networking/nat/vcn.tf:
```diff
@@ -0,0 +1,215 @@
+variable "tenancy_ocid" {}
```
This file should probably be called nat.tf - seems to be the pattern we're using.

briangustafson commented on docs/examples/networking/nat/user_data.tpl:
```diff
@@ -0,0 +1,11 @@
+    content: |
+      net.ipv4.ip_forward = 1
+
+runcmd:
```
Would be useful to add a comment on what this is doing (for those not so familiar with these tools, such as myself).

briangustafson commented on docs/examples/networking/nat/env-vars:
```diff
@@ -0,0 +1,8 @@
+### Region
```
This file should have all of the variables used.

briangustafson commented on docs/examples/networking/nat/env-vars:
```diff
@@ -0,0 +1,8 @@
+### Compartment
+export TF_VAR_compartment_ocid=""
+
+export TF_VAR_ssh_public_key=$(cat .../id_rsa.pub)
```
How about the follow, copied from another example:

export TF_VAR_ssh_public_key=$(cat <path to public key>)
export TF_VAR_ssh_private_key=$(cat <path to private key>)

briangustafson commented on docs/examples/networking/nat/vcn.tf:
```diff
@@ -0,0 +1,215 @@
+
+resource "oci_core_subnet" "MgmtSubnet" {
+    availability_domain = "${lookup(data.oci_identity_availability_domains.ADs.availability_domains[var.AD - 1],"name")}"
+    cidr_block = "10.0.0.0/24"
```
The VCN CIDR has a variable, but the Subnet CIDRs do not. I would think that if one is a variable then the others should be as well.

briangustafson commented on docs/examples/networking/nat/vcn.tf:
```diff
@@ -0,0 +1,215 @@
+data "oci_identity_availability_domains" "ADs" {
+    compartment_id = "${var.tenancy_ocid}"
+}
+variable "VCN-CIDR" {
```
Move to top with other vars.

briangustafson commented on docs/examples/networking/nat/nat.tf:
```diff
@@ -0,0 +1,226 @@
+}
+
+variable "private_subnet_cidr" {
+    default = "10.0.0.0/16"
```
When running using these defaults I'm getting the error below - I think this needs to be a different value from the other subnet:

* oci_core_subnet.PrivateSubnet: Status: 400; Code: InvalidParameter; OPC Request ID: /87AF334F38C76347EE6B866A2F26EA55/943BBE5065234BEE81B795DD3DFC2518; Message: The requested CIDR 10.0.0.0/16 is invalid: subnet ocid1.subnet.oc1.iad.aaaaaaaa4fdarrsa7jyqy7wrs5athfpct55dea2n5cntqypu4grd5fehjjwa with CIDR 10.0.0.0/16 overlaps with this CIDR.

sivaedupuganti-oracle commented:
@briangustafson fixed the errors with variables. Updated the hybrid DNS example to launch two DNS VMs in separate ADs, and added a script to modify default DHCP options.

briangustafson commented on docs/examples/networking/nat/user_data.tpl:
```diff
@@ -0,0 +1,18 @@
+#
```
What's the simplest way to verify that this is working correctly?

sivaedupuganti-oracle replied:
Addressed this in the README.

sivaedupuganti-oracle commented:
Updated hybrid_dns to remove the steps to update default DHCP options and left a note in Readme instead. Moved docx to markdown.

briangustafson commented on docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md:
```diff
@@ -0,0 +1,78 @@
+   * $ sudo firewall-cmd --permanent --add-port=53/udp
+   * $ sudo firewall-cmd --permanent --add-port=53/tcp
+   * $ sudo /bin/systemctl restart firewalld
+   * $ cat &gt; /etc/named.conf
```
The formatting for this does not show up correctly. (The the "view" button for this file in the upper right.)

briangustafson commented on docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md:
```diff
@@ -0,0 +1,78 @@
+# Hybrid DNS Configuration
+
+Oracle Cloud Infrastructure (OCI) customers can configure DNS names for their instances in the Virtual Cloud Network (VCN) as described in [DNS in Your Virtual Cloud Network](https://docs.us-phoenix-1.oraclecloud.com/Content/Network/Concepts/dns.htm). The DNS names are resolvable only within the VCN using the VCN DNS resolver available at 169.254.169.254 IP address. This IP address is only reacha ...
```
"available at IP address 169.254.169.254"

briangustafson commented on docs/examples/networking/nat/README.md:
```diff
@@ -0,0 +1,29 @@
+  * `$ . env-vars`
+* Update variables in vcn.tf as applicable to your target environment.
+
+Once the environment is built, the private instance has Internet connectivity even when it doesn't have a public IP address and it's subnet's route table doesn't contain Internet gateway. You can login into the private instance (from the nat instance) and then run a command like 'ping oracle.com' to verify connectivity.
```
"and its subnet's route table doesn't contain an Internet gateway."

sivaedupuganti-oracle commented:
@briangustafson this PR contains two major scenarios that can be implemented in networking. I'm interested to get this in front of people who want to use this. Let's push this version. Any further feedback can be iterated in a future version, at which time there will be feedback at broader level from other sources too.
